cloud/aws/services/ec2/natgateways.go

```diff
@@ -0,0 +1,110 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ec2
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
+)
+
+func (s *Service) reconcileNatGateways(subnets []*v1alpha1.Subnet, vpc *v1alpha1.VPC) error {
+	if !s.containsPrivateSubnets(subnets) {
+		return nil
+	}
+
+	existing, err := s.describeNatGatewaysBySubnet(vpc.ID)
+	if err != nil {
+		return err
+	}
+
+	for _, sn := range subnets {
+		if sn.ID == "" || !sn.IsPublic {
+			continue
+		}
+		if _, ok := existing[sn.ID]; ok {
+			continue
+		}
+
+		_, err := s.createNatGateway(sn.ID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) describeNatGatewaysBySubnet(vpcID string) (map[string]*ec2.NatGateway, error) {
+	describeNatGatewayInput := ec2.DescribeNatGatewaysInput{
+		Filter: []*ec2.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []*string{aws.String(vpcID)},
+			},
+		},
+	}
+
+	gateways := make(map[string]*ec2.NatGateway)
+
+	err := s.ec2.DescribeNatGatewaysPages(
+		&describeNatGatewayInput,
+		func(page *ec2.DescribeNatGatewaysOutput, lastPage bool) bool {
+			for _, r := range page.NatGateways {
+				gateways[*r.SubnetId] = r
+			}
+			return false
+		})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to describe NAT gateways with VPC ID %q", vpcID)
+	}
+	return gateways, nil
+}
+
+func (s *Service) createNatGateway(subnetID string) (*ec2.NatGateway, error) {
+
+	ip, err := s.allocateAddress()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create IP address for NAT gateway for subnet ID %q", subnetID)
+	}
+
+	out, err := s.ec2.CreateNatGateway(&ec2.CreateNatGatewayInput{
+		SubnetId:     aws.String(subnetID),
+		AllocationId: aws.String(ip),
+	})
+
+	{
+		err := s.ec2.WaitUntilNatGatewayAvailable(&ec2.DescribeNatGatewaysInput{
+			NatGatewayIds: []*string{out.NatGateway.NatGatewayId},
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create NAT gateway for subnet ID %q", subnetID)
+		}
+	}
+
+	return out.NatGateway, nil
+}
+
+func (s *Service) findNatGatewayByID(gatewayID string) (*ec2.NatGateway, error) {
+	out, err := s.ec2.DescribeNatGateways(&ec2.DescribeNatGatewaysInput{
+		NatGatewayIds: []*string{&gatewayID},
+	})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to describe NAT gateway with ID %q", gatewayID)
+	}
+
+	return out.NatGateways[0], nil
```

Comment:
I believe we'd need to check bounds here, and return a `NotFound` error in case the slice is empty. See https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/cloud/aws/services/ec2/gateways.go#L71-L73

Reply:
Thanks. Added.